proposals/consolidate-merged-lifecycle-hook-schema.md

```diff
@@ -0,0 +1,141 @@
+
+As an example, the following `devcontainer.json` snippet would be valid:
+
+```json
+{
+	"onCreateCommand": [
+		{
+			"commandA": "echo 'Hello World!'",
+			"commandB": ["echo", "'Hello World!'"],
+			"$origin": "featureA"
+		},
+		{
+			"commandA": "echo 'Hello World!'",
+			"commandB": ["echo", "'Hello World!'"],
+			"$origin": "devcontainer.json"
+		}
+	]
+}
+```
+
+#### A2. Extend `customizations` property
+
+The merging of the `customizations` property is left to the implementing tool, therefore there is no defined merging pattern. 
+
+Extend the interface of to include another unioned type:
+
+```typescript
+interface DevContainerConfig {
+	// ...other devcontainer.json properties...
+	customizations?: Record<string, {}> | Record<string, any[]>
+}
+```
+
+The addition of union type `Record<string, any[]>` (and explicitly typing the existing case to an object) allows for various tools to contribute customizations to the same namespace.  For example, a Feature may contribute a `settings` object, and a user (via `devcontainer.json`) may contribute a `settings` object.  The merged `customizations` property would be an array of two objects.  This allows for the merging algorithm to be defined by the implementing tool.
+
+Similar to above, the `$origin` property may be added by tools to indicate the source of the customization.
+
+In the below example, the `vscode` customization has two entries (contributed from two different sources), and the `foo` customization has two entries (contributed from two different sources).  The merged `customizations` property would be an array of two objects for each customization namespace.
+
+```json
+"customizations": {
+	// Customizations for the 'vscode' namespace
+	"vscode": [
+		{
+			"settings": {
+				"settingA": "local",
+				"settingB": "/usr/bin/lldb"
+			},
+			"extensions": [
+				"GitHub.vscode-pull-request-github"
+			],
+			"$origin": "featureA"
+		},
+		{
+			"settings": {
+				"settingA": "local",
+				"settingC": true
+			},
+			"$origin": "devcontainer.json"
+		}
+	],
+	// Customizations for the 'foo' namespace
+	"foo": [
+		{
+			"bar": "baz",
+			"b": true,
+			"$origin": "featureA"
+		},
+		{
+			"bar": "baz",
+			"a": true,
+			"$origin": "devcontainer.json"
+		}
+	]
+}
+...
+```
+
+### B. Generating a merged configuration
+
+The [reference implementation](https://github.com/devcontainers/cli) includes a 'read-configuration' command that can be used to resolve a fully merged configuration from a given `devcontainer.json`.
+
+Implementing tools should follow the [documented merging logic](https://containers.dev/implementors/spec/#merge-logic) and output a `devcontainer.json` that represents the resolved configuration.
```

Comment:
Maybe omit this paragraph ("Implementing tools should...") as it may not make sense for every implementing tool to offer such a command (implementers might not write a CLI at all).